CLOUDNS: NS Record pointing to @ integration test failing

```
# go test -v -verbose -provider CLOUDNS -start 15 -end 15
=== RUN   TestDNSProviders
=== RUN   TestDNSProviders/example.com
=== RUN   TestDNSProviders/example.com/Clean_Slate:Empty
=== RUN   TestDNSProviders/example.com/15:NS:NS_for_subdomain
    integration_test.go:223:
        + CREATE NS xyz.example.com ns2.foo.com. ttl=300
=== RUN   TestDNSProviders/example.com/15:NS:Dual_NS_for_subdomain
    integration_test.go:223:
        + CREATE NS xyz.example.com ns1.foo.com. ttl=300
=== RUN   TestDNSProviders/example.com/15:NS:NS_Record_pointing_to_@
    integration_test.go:223:
        - DELETE NS xyz.example.com ns1.foo.com. ttl=300, ClouDNS ID: 283378001
    integration_test.go:223:
        - DELETE NS xyz.example.com ns2.foo.com. ttl=300, ClouDNS ID: 283377999
    integration_test.go:223:
        + CREATE NS foo.example.com example.com. ttl=300
    integration_test.go:227: failed create record (ClouDNS): ClouDNS API error: The NS record needs to be pointed to a hostname that has an A or an AAAA record. URL:api.cloudns.net/dns/add-record.json?auth-id=**REDACTED**&auth-password=**REDACTED**&domain-name=example.com&host=foo&record=example.com&record-type=NS&sub-auth-id=&ttl=300
=== RUN   TestDNSProviders/example.com/Post_cleanup:Empty
--- FAIL: TestDNSProviders (2.62s)
    --- FAIL: TestDNSProviders/example.com (2.31s)
        --- PASS: TestDNSProviders/example.com/Clean_Slate:Empty (0.23s)
        --- PASS: TestDNSProviders/example.com/15:NS:NS_for_subdomain (0.58s)
        --- PASS: TestDNSProviders/example.com/15:NS:Dual_NS_for_subdomain (0.57s)
        --- FAIL: TestDNSProviders/example.com/15:NS:NS_Record_pointing_to_@ (0.58s)
        --- PASS: TestDNSProviders/example.com/Post_cleanup:Empty (0.23s)
=== RUN   TestDualProviders
    integration_test.go:318: Skipping.  DocDualHost == Cannot
--- SKIP: TestDualProviders (0.26s)
FAIL
exit status 1
FAIL    github.com/StackExchange/dnscontrol/v3/integrationTest  2.887s
```

issue root

https://github.com/StackExchange/dnscontrol/blob/f700184e3803c9b60cf43dc5960155a39da721cf/providers/cloudns/cloudnsProvider.go#L158-L165

As I can see here, A/AAAA records need to be created before NS, NS records need to be created before DS

Self-assigned, will create a PR soon.

--- a/providers/cloudns/auditrecords.go
+++ b/providers/cloudns/auditrecords.go
@@ -21,5 +21,7 @@ func AuditRecords(records []*models.RecordConfig) []error {
 
 	a.Add("TXT", rejectif.TxtHasMultipleSegments) // Last verified 2021-03-01
 
+	a.Add("SRV", rejectif.SrvHasNullTarget) // Last verified 2023-03-30
+
 	return a.Audit(records)
 }

--- a/providers/cloudns/cloudnsProvider.go
+++ b/providers/cloudns/cloudnsProvider.go
@@ -136,7 +136,11 @@ func (c *cloudnsProvider) GetDomainCorrections(dc *models.DomainConfig) ([]*mode
 		}
 	}
 
-	var createCorrections []*models.Correction
+	var (
+		createCorrections         []*models.Correction
+		createARecordCorrections  []*models.Correction
+		createNSRecordCorrections []*models.Correction
+	)
 	for _, m := range create {
 		req, err := toReq(m.Desired)
 		if err != nil {
@@ -155,15 +159,20 @@ func (c *cloudnsProvider) GetDomainCorrections(dc *models.DomainConfig) ([]*mode
 				return c.createRecord(domainID, req)
 			},
 		}
-		// at ClouDNS, we MUST have a NS for a DS
-		// So, when creating, we must create the NS first, otherwise creating the DS throws an error
-		if m.Desired.Type == "NS" {
-			// type NS is prepended - so executed first
-			createCorrections = append([]*models.Correction{corr}, createCorrections...)
-		} else {
+		// A & AAAA need to be created before NS #2244
+		// NS need to be created before DS #1018
+		// or else errors will be thrown
+		switch m.Desired.Type {
+		case "A", "AAAA":
+			createARecordCorrections = append(createARecordCorrections, corr)
+		case "NS":
+			createNSRecordCorrections = append(createNSRecordCorrections, corr)
+		default:
 			createCorrections = append(createCorrections, corr)
 		}
 	}
+	corrections = append(corrections, createARecordCorrections...)
+	corrections = append(corrections, createNSRecordCorrections...)
 	corrections = append(corrections, createCorrections...)
 
 	for _, m := range modify {

